### DISCUSSION TITLE
3rd Party Controls Wishlist

### DISCUSSION THREAD

**Discussion by PDBNZ** [twinbasic/twinbasic#1470] [Ideas]
Did not see one listed so thought I would get my wishlist out there for any 3rd Party developers creating libraries/controls for twinBasic.
My preference is for these libraries/controls to be included via packages (I think that is the term) and not as an Active X control.
I want a self-contained single exe to distribute.

In no particular order:

1. Ribbon control.
2. Grid with all the bells and whistles.
3. Chart control.
4. Reporting component.
5. Pivot/OLAP control.

Thank you.

**Comment by bclothier** [comment]
Just in case... it may be good to know that the controls used by tB themselves are written in tB and you can inspect the source code behind the common controls used by the tB and therefore use them as an inspiration for developing custom controls. [Fafalone also has some great samples that was ported to tB as well](https://github.com/fafalone).

It's very likely that those will come from the community and in form of a twinpack package rather than a native control. Indeed, I'd argue if we want tB to succeed, the community will have to contribute such packages and thus free up Wayne to work on hard problems deep under the hood.

**Comment by wqweto** [comment]
For reporting TB must first support ActiveX Designers like DataReport or 3-rd party ones like ActiveReports which is not on the roadmap for v1 or even later ones AFAIK. 

Probably some other kind of reporting will emerge from this, based on HTML or PDFs, or we might get some “native” comparible framework from TB instead of Ax Designers.

**Reply by GaryMiller** [reply]
For AI and Chatbot Technology
1. Cross Platform Speech Generation Control
2. In Memory Graph Database Control
3. Messaging Control for MicroService applications.

**Comment by fafalone** [comment]
There's already some great grid and ribbon controls for VB6, these should mostly work unmodified in tB in x86 mode at least, or with a minor rework to swap out the assembly thunk based and other incompatible self-subclassing methods you find in a lot of UCs. 

If they're popular enough, most are released under licenses that would allow us to make an x64-compatible port if the author isn't around anymore (or is and gives permission). It's not really that difficult at all, having done both this and swapping subclassing for my own large, complex controls, just a bit time consuming.
